### PR TITLE
Align retry logic for all test framework instrumentations

### PR DESCRIPTION
**Pull request by daniel-mohedano**
# What Does This Do

This PR aligns all the retry logic to be used in the same manner between all framework instrumentations. The logic flow for retries after the refactoring is:
1. Check if the execution policy is applicable. This is normally only done once before the first test execution to avoid performance overhead if the policy is not going to be applicable for the test.
2. Check if the failure for the following test execution should be suppressed or not. In the frameworks where this is used after test execution, the value is stored for later use.
3. Register test execution in the history, currently performed in `datadog.trace.civisibility.events.TestEventsHandlerImpl#onTestFinish`. This will update the policy logic with the information about the test run to inform later retries.
4. Check if the test should be retried with `datadog.trace.api.civisibility.execution.TestExecutionHistory#wasLastExecution`. In cases where this is performed by the testing framework before reporting the test result, the test result is manually reported in order to register its execution beforehand.  

Incidentally, this refactoring also fixes another issue where `TestNG` was marking all retried tests as failed, independently of their result.

# Motivation

There were several issues with the previous implementation of the retry logic:
- The registration of test results in the history was performed after test reporting. EFD can change the maximum number of retries, creating certain edge cases where a test execution was not reported as the last one but was considered as such after the fact. 
- The `TestNG` framework had an ordering disparity with the rest of the frameworks, marking as last execution actually the second to last execution.

# Additional Notes

# Contributor Checklist

- Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

**Review by nikita-tkachenko-datadog — COMMENTED** [review]
Nice, I like that it actually simplify the instrumentations (apart from Test NG), I was not expecting that!
Do I understand correctly, that Test NG was the only framework where we found problems, which is why test fixtures for the other instrumentations are not being updated?

File: dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/domain/AbstractTestModule.java [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import static datadog.trace.civisibility.Constants.CI_VISIBILITY_INSTRUMENTATION_NAME;
 
 import datadog.trace.api.Config;
+import datadog.trace.api.civisibility.execution.TestStatus;
```
[the remark this answers is not in the document]
Comment (daniel-mohedano) [reply dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/domain/AbstractTestModule.java:6]: The line change is due to a refactor. I moved the `TestStatus` class to the API package, previously it was only internal

File: dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/execution/RetryUntilSuccessful.java [hunk 2]
```diff
@@ -25,30 +26,33 @@ public RetryUntilSuccessful(
   }
 
   @Override
-  public boolean applicable() {
-    return !currentExecutionIsLast() || suppressFailures;
+  public void registerExecution(TestStatus status, long durationMillis) {
+    ++executions;
+    successfulExecutionSeen |= (status != TestStatus.fail);
+    if (executions > 1) {
```
[the remark this answers is not in the document]
Comment (daniel-mohedano) [reply dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/execution/RetryUntilSuccessful.java:32]: From my understanding of `datadog.trace.civisibility.domain.buildsystem.ProxyTestModuleTest` (and the equivalent test for headless modules), the global counter for ATR retries only takes into account actual retries, so we want to avoid increasing it after the first execution of a test. Should it take into account the first execution too?

File: dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/execution/RunNTimes.java [hunk 3]
```diff
@@ -25,20 +26,38 @@ public RunNTimes(
...
+
+  @Override
+  public boolean wasLastExecution() {
+    // skipped tests (either by the framework or DD) should not be retried
+    return lastStatus == TestStatus.skip || executions >= maxExecutions;
   }
 
   private boolean currentExecutionIsLast() {
+    // this could give false negatives if maxExecutions is updated after the current execution is
+    // registered
```
[the remark this answers is not in the document]
Comment (daniel-mohedano) [reply dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/execution/RunNTimes.java:52]: The issue is the order in which the check is done. Because this is performed before the execution is registered, we don't have a way of knowing if EFD will update the number of maximum retries, decreasing it under the value of `executions` (and therefore making it a false negative after the fact)

File: dd-java-agent/instrumentation/scalatest/src/main/java/datadog/trace/instrumentation/scalatest/RunContext.java [hunk 4]
```diff
@@ -162,18 +162,6 @@ public TestExecutionHistory getExecutionHistory(TestIdentifier testIdentifier) {
     return executionPolicies.get(testIdentifier);
   }
 
-  @Nullable
-  public TestExecutionHistory popExecutionHistory(TestIdentifier testIdentifier) {
```
[the remark this answers is not in the document]
Comment (daniel-mohedano) [reply dd-java-agent/instrumentation/scalatest/src/main/java/datadog/trace/instrumentation/scalatest/RunContext.java:166]: Ah, I didn't fully understand how the method was useful, as it was the only framework in which the `applicable` method was being called repeatedly for every test execution. But we wouldn't have any issue with adding it back, taking into account the changes to considering a policy as applicable I mentioned in the previous comment

File: dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/execution/RetryUntilSuccessful.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply dd-java-agent/agent-ci-visibility/src/main/java/datadog/trace/civisibility/execution/RetryUntilSuccessful.java:32]: Makes sense. I'd say let's rename `totalExecutions` to `totalRetries` then and maybe leave a short comment

**Review by nikita-tkachenko-datadog — APPROVED** [review]
